### ISSUE TITLE
Feature: More typed API

### ISSUE THREAD

**Issue by DDEle** [bitomic/mwjs#161]
Hi there, I'm switching from `mwbot` to this repo for typing support, However, I think there are some queries that currently I'm cannot find the equivalents:
``` javascript
bot.request({
    action: 'query',
    formatversion: 2,
    iwurl: 1,
    titles: title,
    redirects: 1,
    converttitles: 1,
    prop: 'info',
    inprop: 'url|displaytitle',
})
```
``` javascript
bot.request({
      action: 'opensearch',
      format: 'json',
      search,
      redirects: 'resolve',
      limit: 3,
})
```
``` javascript
bot.request({
      action: 'parse',
      pageid,
      prop: 'text|wikitext',
      wrapoutputclass: 'mw-parser-output',
      disablelimitreport: 1,
      disableeditsection: 1,
      disabletoc: 1,
})
```

It would also be appreciated if the `parsedcomment` in `reprop` of `recentchanges` can be supported, while I noticed that it is already in the dev branch of this repo.

**Comment by bitomic** [comment]
The `refactor` branch was merged in #137. Looks like I just forgot about the `parsedcomment` prop (and also forgot to delete the branch).

- Add `parsedcomment` to the options for `rcprop` (8b8309b)
- Add `Wiki#search` (using API:Opensearch) (001b386)
- Add `Wiki#parse` (using API:Parse) (fabe2d1)

Before I can merge these changes, I noticed in your example:
```ts
bot.request({
    action: 'query',
    formatversion: 2,
    iwurl: 1,
    titles: title,
    redirects: 1,
    converttitles: 1,
    prop: 'info',
    inprop: 'url|displaytitle',
})
```
[mw:API:Info](www.mediawiki.org/wiki/API:Info) doesn't seem to mention `iwurl`, `redirects` nor `converttitles`. Did I miss something?

----
On a side note, thanks for giving `mw.js` a chance! And also for helping to improve it through your feature suggestions. ^^

**Comment by DDEle** [comment]
Thanks again for your super fast reaction! For `iwurl`, `redirects` and `converttitles`, they are parameters for [`action=query`](https://www.mediawiki.org/wiki/API:Query)

----
It's my luck to find this library of typed Mediawiki API. And I really appreciate your active development! :)

**Comment by bitomic** [comment]
Absolutely feel free to PR any small change you need. I don't really have an ongoing plan but just add stuff on the go as I need them. That's why there aren't all `list=` or `prop=` typings available, even though it would be quite straight-forward to implement in most cases, if not all (just by adding its interfaces —[request](https://github.com/bitomic/mwjs/blob/d8bde44a16c9145fd787d1e5c4d43d1bccc39ae5/src/types.ts#L594-L646) and [response](https://github.com/bitomic/mwjs/blob/d8bde44a16c9145fd787d1e5c4d43d1bccc39ae5/src/types.ts#L1431-L1437)— and overloading a couple of `Wiki`'s methods —[query](https://github.com/bitomic/mwjs/blob/d8bde44a16c9145fd787d1e5c4d43d1bccc39ae5/src/mediawiki/_base/Wiki.ts#L226-L234) and [iterQuery](https://github.com/bitomic/mwjs/blob/d8bde44a16c9145fd787d1e5c4d43d1bccc39ae5/src/mediawiki/_base/Wiki.ts#L243-L252)—).

Either make a PR or an issue, both are fine!

- Added `action=query` parameters (4797b2c)

Merged #163 and published as `mw.js@2.3.0`.

**Comment by DDEle** [comment]
Thanks for the update!

**Comment by DDEle** [comment]
Well, it seems that supporting redirects is more complex than we thought.
The problem is that we assumed that the query will only have "pages" filed (currently the return type of `queryProp` is only for pages), while it can have `interwiki`, `redirects`, `normalized` etc (simultaneously or independently). Here is [an example of that](https://oxygennotincluded.fandom.com/wiki/Special:ApiSandbox#action=query&format=json&prop=&iwurl=1&titles=Project%3ACommunity%20portal%7C211197%7Czh%3A%E6%B6%B2%E4%BD%93&redirects=1&formatversion=2).

**Comment by bitomic** [comment]
I think it may be easier to have an, e.g., `rawQueryProp` method that returns the full object instead of filtering just the pages (especially to avoid a breaking change to `queryProp`/`queryList`). I will be adding it tomorrow that I am free. Of course, you can PR it if you prefer; besides just adding the method, including the `interwiki`, `redirects` and `normalized` properties in `QueryResponse` should work fine.

**Comment by DDEle** [comment]
OK, I can create a PR for this.

**Comment by bitomic** [comment]
Closed by #173
